### PR TITLE
feat: handle webhook after stripe payment

### PR DESCRIPTION
**Pull request by vackca**

### CODE REVIEW

File: microservices/payment/src/entities/transaction.ts [hunk 1]
```diff
@@ -1,6 +1,54 @@
...
+  properties: {
+    customer: { $ref: '#/definitions/Customer' },
+  },
+})
 @Entity()
-class Transaction {}
+class Transaction {
+  @PrimaryColumn({ type: 'varchar', length: 66 })
+  @Allow()
+  transactionId: string;
```
Comment (MatthewPattell) [inline microservices/payment/src/entities/transaction.ts:30]:
```suggestion
  id: string;
```

File: microservices/payment/src/entities/customer.ts [hunk 2]
```diff
@@ -1,20 +1,26 @@
...
+})
 @Entity()
 class Customer {
   @PrimaryColumn({ type: 'varchar', length: 18 })
   @Allow()
   customerId: string;
 
   @Index('IDX_payment_userId', ['userId'])
-  @PrimaryColumn({ type: 'varchar', length: 36 })
+  @Column({ type: 'varchar', length: 36 })
```
Comment (MatthewPattell) [inline microservices/payment/src/entities/customer.ts:23]: revert

**Review by MatthewPattell — CHANGES_REQUESTED** [review]

File: microservices/payment/src/entities/transaction.ts [hunk 3]
```diff
@@ -1,6 +1,74 @@
...
+  @IsUndefinable()
+  cardId: string;
+
+  @Column({ type: 'varchar', length: 32 })
+  @IsUndefinable()
+  entityId: string;
+
+  @Column({ type: 'int' })
+  @Allow()
+  amount: number;
```
Comment (MatthewPattell) [inline microservices/payment/src/entities/transaction.ts]: it's can be e.g. 12.45 also, so int not correct

File: microservices/payment/src/entities/transaction.ts [hunk 4]
```diff
@@ -1,6 +1,74 @@
...
+  @IsUndefinable()
+  type: string;
+
+  @Column({ type: 'int' })
+  @IsUndefinable()
+  tax: number;
+
+  @Column({ type: 'int' })
+  @IsUndefinable()
+  fee: number;
```
Comment (MatthewPattell) [inline microservices/payment/src/entities/transaction.ts]: the same

File: microservices/payment/src/entities/transaction.ts [hunk 5]
```diff
@@ -1,6 +1,74 @@
...
+  @IsUndefinable()
+  entityId: string;
+
+  @Column({ type: 'int' })
+  @Allow()
+  amount: number;
+
+  @Column({ type: 'varchar', length: 10 })
+  @IsUndefinable()
+  type: string;
```
Comment (MatthewPattell) [inline microservices/payment/src/entities/transaction.ts]: enum?

File: microservices/payment/src/entities/transaction.ts [hunk 6]
```diff
@@ -1,6 +1,74 @@
...
+  properties: {
+    customer: { $ref: '#/definitions/Customer' },
+  },
+})
 @Entity()
-class Transaction {}
+class Transaction {
+  @PrimaryColumn({ type: 'varchar', length: 66 })
+  @Allow()
+  id: string;
```
Comment (MatthewPattell) [inline microservices/payment/src/entities/transaction.ts]: why it not auto increment? uuid? (if it's external transaction id, you should store it in separated field

File: microservices/payment/src/services/payment-gateway/abstract.ts [hunk 7]
```diff
@@ -23,6 +24,21 @@ export interface IPriceParams {
...
 
+export interface ITransactionParams {
+  id: string;
+  title?: string;
+  amount: number;
+  customerId: string;
+  bankAccountId?: string;
+  cardId?: string;
+  entityId?: string;
+  type?: string;
```
Comment (MatthewPattell) [inline microservices/payment/src/services/payment-gateway/abstract.ts]: enum

File: microservices/payment/src/services/payment-gateway/abstract.ts [hunk 8]
```diff
@@ -23,6 +24,21 @@ export interface IPriceParams {
...
+  tax?: number;
+  fee?: number;
+  status: string;
```
Comment (MatthewPattell) [inline microservices/payment/src/services/payment-gateway/abstract.ts]: enum?

File: microservices/payment/src/services/payment-gateway/abstract.ts [hunk 9]
```diff
@@ -82,6 +104,45 @@ abstract class Abstract {
...
+      status,
+      type,
+      bankAccountId,
+      cardId,
+      entityId,
+      amount,
+      tax,
+      fee,
+      customerId,
+    });
```
Comment (MatthewPattell) [inline microservices/payment/src/services/payment-gateway/abstract.ts]:
```suggestion
    const transaction = this.transactionRepository.create(params);
```

???

File: microservices/payment/src/services/payment-gateway/stripe.ts [hunk 10]
```diff
@@ -189,6 +204,41 @@ class Stripe extends Abstract {
...
+    try {
+      const event = this.paymentEntity.webhooks.constructEvent(payload, signature, webhookKey);
+
+      switch (event.type) {
+        case 'checkout.session.completed':
+          void this.handleTransactionCompleted(event);
+          break;
+      }
+    } catch (err) {
+      console.log('Handling webhook ended with the error', err);
```
Comment (MatthewPattell) [inline microservices/payment/src/services/payment-gateway/stripe.ts]: use logger

**Review by MatthewPattell — CHANGES_REQUESTED** [review]

File: microservices/payment/src/entities/transaction.ts (hunk 3 above)
Comment (vackca) [reply microservices/payment/src/entities/transaction.ts]: Stripe uses integers for prices like 12.5 $ is equal 1250 (cents). I checked paypal, paypal uses even strings. Can I leave integers for now?

Comment (vackca) [reply microservices/payment/src/entities/transaction.ts]: image

File: microservices/payment/src/entities/transaction.ts (hunk 4 above)
Comment (vackca) [reply microservices/payment/src/entities/transaction.ts]: answered above

File: microservices/payment/src/entities/transaction.ts (hunk 6 above)
Comment (vackca) [reply microservices/payment/src/entities/transaction.ts]: yes, this is external transaction id. The first version of this field I named was trnasactionId. Is transactionId correct?

**Comment by vackca** [comment]
deleted all changes related to gateway ms and created separate pull request

File: microservices/payment/src/entities/transaction.ts (hunk 3 above)
Comment (MatthewPattell) [reply microservices/payment/src/entities/transaction.ts]: ok, for all integrations will be used as int

File: microservices/payment/src/entities/transaction.ts [hunk 11]
```diff
@@ -1,6 +1,89 @@
...
+  },
+})
 @Entity()
-class Transaction {}
+class Transaction {
+  @PrimaryGeneratedColumn('uuid')
+  @Allow()
+  id: string;
+
+  @PrimaryColumn({ type: 'varchar', length: 66 })
```
Comment (MatthewPattell) [inline microservices/payment/src/entities/transaction.ts]: you can use `@Column`

File: microservices/payment/src/constants/payment-intent-status.ts [hunk 12]
```diff
@@ -0,0 +1,10 @@
+/**
+ * Enum for storing payment intent status
+ */
+enum PaymentIntentStatus {
```
Comment (MatthewPattell) [inline microservices/payment/src/constants/payment-intent-status.ts]:
```suggestion
enum TransactionStatus {
```

File: microservices/payment/src/constants/payment-intent-status.ts [hunk 13]
```diff
@@ -0,0 +1,10 @@
+/**
+ * Enum for storing payment intent status
```
Comment (MatthewPattell) [inline microservices/payment/src/constants/payment-intent-status.ts]:
```suggestion
 * Enum for storing transaction status
```

File: microservices/payment/src/constants/payment-intent-status.ts (hunk 12 above)
Comment (MatthewPattell) [reply microservices/payment/src/constants/payment-intent-status.ts]: add `error` status

Comment (MatthewPattell) [reply microservices/payment/src/constants/payment-intent-status.ts]: also we need store error message somewhere (add column to Transaction? e.g. errorMessage)

**Review by MatthewPattell — CHANGES_REQUESTED** [review]
Update migration after all changes!

File: microservices/authorization/migrations/permissions/list/models/payment.json [hunk 14]
```diff
@@ -555,6 +567,191 @@
     },
     "createdAt": "2023-04-19T09:58:19.118Z"
   },
+  {
+    "microservice": "payment",
+    "alias": "payment.Transactions",
+    "title": "Transactions",
+    "schema": {},
+    "createdAt": "2023-04-25T14:18:01.068Z"
+  },
```
Comment (MatthewPattell) [inline microservices/authorization/migrations/permissions/list/models/payment.json]: ?

Comment (MatthewPattell) [reply microservices/authorization/migrations/permissions/list/models/payment.json]: review all models

File: microservices/payment/src/constants/transaction-status.ts [hunk 15]
```diff
@@ -0,0 +1,11 @@
+/**
+ * Enum for storing transaction status
+ */
+enum TransactionStatus {
+  SUCCEDED = 'succeded',
+  REQUIRES_ACTION = 'requires_action',
+  REQUIRES_PAYMENT_METHOD = 'requires_payment_method',
```
Comment (MatthewPattell) [inline microservices/payment/src/constants/transaction-status.ts]:
```suggestion
  SUCCEDED = 'succeded',
  REQUIRES_ACTION = 'requiresAction',
  REQUIRES_PAYMENT_METHOD = 'requiresPaymentMethod',
```

File: microservices/payment/src/entities/customer.ts [hunk 16]
```diff
@@ -1,20 +1,27 @@
...
   accountId?: string;
 }
 
+@JSONSchema({
+  properties: {
+    transactions: { $ref: '#/definitions/Transactions', type: 'array' },
+  },
+})
 @Entity()
+@Unique(['userId'])
```
Comment (MatthewPattell) [inline microservices/payment/src/entities/customer.ts]: move to column

File: microservices/payment/src/entities/transaction.ts [hunk 17]
```diff
@@ -1,6 +1,86 @@
...
+    description:
+      'Field for storing status of process of transaction itself. E.g. checkout session is completed or in process',
+  })
+  @Column({ type: 'varchar', length: 18 })
+  @IsUndefinable()
+  transactionStatus: string;
+
+  @Column({ type: 'varchar', length: 150 })
+  @IsUndefinable()
+  errorMessage: string;
```
Comment (MatthewPattell) [inline microservices/payment/src/entities/transaction.ts]: add validator

File: microservices/payment/src/entities/transaction.ts [hunk 18]
```diff
@@ -1,6 +1,86 @@
...
+  @IsUndefinable()
+  status: TransactionStatus;
+
+  @JSONSchema({
+    description:
+      'Field for storing status of process of transaction itself. E.g. checkout session is completed or in process',
+  })
+  @Column({ type: 'varchar', length: 18 })
+  @IsUndefinable()
+  transactionStatus: string;
```
Comment (MatthewPattell) [inline microservices/payment/src/entities/transaction.ts]: add validator

File: microservices/payment/src/entities/transaction.ts [hunk 19]
```diff
@@ -1,6 +1,86 @@
...
+  @IsEnum(TransactionType)
+  @IsUndefinable()
+  type: TransactionType;
+
+  @Column({ type: 'int' })
+  @IsUndefinable()
+  tax: number;
+
+  @Column({ type: 'int' })
+  @IsUndefinable()
```
Comment (MatthewPattell) [inline microservices/payment/src/entities/transaction.ts]: the same

File: microservices/payment/src/entities/transaction.ts [hunk 20]
```diff
@@ -1,6 +1,86 @@
...
+  cardId: string;
+
+  @Column({ type: 'varchar', length: 32 })
+  @IsUndefinable()
+  entityId: string;
+
+  @Column({ type: 'int' })
+  @Allow()
+  amount: number;
```
Comment (MatthewPattell) [inline microservices/payment/src/entities/transaction.ts]: add normal validator, not `@Allow()`

File: microservices/payment/src/entities/transaction.ts [hunk 21]
```diff
@@ -1,6 +1,86 @@
...
+import Customer from '@entities/customer';
 
+@JSONSchema({
+  properties: {
+    customer: { $ref: '#/definitions/Customer' },
+  },
+})
 @Entity()
-class Transaction {}
+class Transaction {
```
Comment (MatthewPattell) [inline microservices/payment/src/entities/transaction.ts:23]: add validators for all fields

File: microservices/payment/src/services/payment-gateway/abstract.ts [hunk 22]
```diff
@@ -23,6 +26,21 @@ export interface IPriceParams {
...
+  userId: string;
+  bankAccountId?: string;
+  cardId?: string;
+  entityId?: string;
+  type?: TransactionType;
+  tax?: number;
+  fee?: number;
+  errorMessage?: string;
+  status: TransactionStatus;
+  transactionStatus: string;
```
Comment (MatthewPattell) [inline microservices/payment/src/services/payment-gateway/abstract.ts]: looks strange

File: microservices/payment/src/services/payment-gateway/abstract.ts [hunk 23]
```diff
@@ -82,11 +106,22 @@ abstract class Abstract {
...
+
+    await this.transactionRepository.save(transaction);
+
+    return transaction;
+  }
+
   /**
    * Get the customer
    */
   protected async getCustomer(userId: string) {
```
Comment (MatthewPattell) [inline microservices/payment/src/services/payment-gateway/abstract.ts]: return type

File: microservices/payment/src/services/payment-gateway/abstract.ts [hunk 24]
```diff
@@ -82,11 +106,22 @@ abstract class Abstract {
    */
   public abstract addBankAccount(params: IBankAccountParams): Promise<BankAccount>;
 
+  /**
+   * Create new transaction
+   */
+  public async createTransaction(params: ITransactionParams, transactionId = uuid()) {
```
Comment (MatthewPattell) [inline microservices/payment/src/services/payment-gateway/abstract.ts]: return type

**Review by MatthewPattell — CHANGES_REQUESTED** [review]

File: microservices/payment/src/services/payment-gateway/abstract.ts (hunk 22 above)
Comment (MatthewPattell) [reply microservices/payment/src/services/payment-gateway/abstract.ts]: we need play with it

Comment (MatthewPattell) [reply microservices/payment/src/services/payment-gateway/abstract.ts]: payment status / status 
for e.g.

File: microservices/payment/src/constants/transaction-status.ts (hunk 15 above)
Comment (vackca) [reply microservices/payment/src/constants/transaction-status.ts]: These are strings that coming from stripe
image

Comment (MatthewPattell) [reply microservices/payment/src/constants/transaction-status.ts]: I know, we need handle it. discuss

File: microservices/payment/src/entities/customer.ts [hunk 25]
```diff
@@ -1,31 +1,42 @@
...
+  properties: {
+    transactions: { $ref: '#/definitions/Transactions', type: 'array' },
+  },
+})
 @Entity()
 class Customer {
   @PrimaryColumn({ type: 'varchar', length: 18 })
-  @Allow()
+  @IsString()
+  @Length(1, 18)
```
Comment (MatthewPattell) [inline microservices/payment/src/entities/customer.ts]:
```suggestion
  @Length(1, 18)
```

Length - enough

fix other fields

File: microservices/payment/src/entities/price.ts [hunk 26]
```diff
@@ -12,30 +12,34 @@ import Product from '@entities/product';
 @Entity()
 class Price {
   @PrimaryColumn({ type: 'varchar', length: 30 })
-  @Allow()
+  @IsString()
```
Comment (MatthewPattell) [inline microservices/payment/src/entities/price.ts]: the same

File: microservices/payment/src/entities/product.ts [hunk 27]
```diff
@@ -16,13 +16,15 @@ class Product {
     description: 'Field for storing id of according product entity created on payment service side',
   })
   @PrimaryColumn({ type: 'varchar', length: 19 })
-  @Allow()
+  @IsString()
```
Comment (MatthewPattell) [inline microservices/payment/src/entities/product.ts]: the same

File: microservices/payment/src/entities/transaction.ts [hunk 28]
```diff
@@ -1,6 +1,102 @@
...
+  @PrimaryGeneratedColumn('uuid')
+  @Allow()
+  id: string;
+
+  @Column({ type: 'varchar', length: 66 })
+  @IsString()
```
Comment (MatthewPattell) [inline microservices/payment/src/entities/transaction.ts]: the same

**Review by MatthewPattell — CHANGES_REQUESTED** [review]

File: microservices/payment/migrations/1679778151988-init.ts [hunk 29]
```diff
@@ -5,18 +5,34 @@ export default class init1679778151988 implements MigrationInterface {
...
+      `CREATE TABLE "product" ("productId" character varying(19) NOT NULL, "entityId" character varying(36) NOT NULL, "userId" character varying(36), CONSTRAINT "product(pk):productId" PRIMARY KEY ("productId"))`,
     );
     await queryRunner.query(
       `CREATE TABLE "price" ("priceId" character varying(30) NOT NULL, "productId" character varying(19) NOT NULL, "userId" character varying(36), "currency" character varying(10) NOT NULL, "unitAmount" integer NOT NULL, CONSTRAINT "price(rel):productId" UNIQUE ("productId"), CONSTRAINT "price(pk):priceId" PRIMARY KEY ("priceId"))`,
     );
+    await queryRunner.query(
+      `CREATE TYPE "public"."transaction_type_enum" AS ENUM('credit', 'debit')`,
+    );
+    await queryRunner.query(
+      `CREATE TYPE "public"."transaction_status_enum" AS ENUM('succeded', 'requires_action', 'requires_payment_method', 'error')`,
```
Comment (MatthewPattell) [inline microservices/payment/migrations/1679778151988-init.ts]: ?????

**Review by MatthewPattell — CHANGES_REQUESTED** [review]

File: microservices/payment/migrations/1679778151988-init.ts (hunk 29 above)
Comment (vackca) [reply microservices/payment/migrations/1679778151988-init.ts]: my bad, forgot, fixed

**Review by MatthewPattell — APPROVED** [review]

**Pull request merged**